File: 2023/Arrays/productOfArrayExceptSelf.cs
class ProductOfArrayExceptSelf {
    static int[] productOfArrayExceptSelf(int[] nums) {
        int[] ans = new int[nums.Length];
        int prefix, suffix;
        prefix = 1;
        for (int i=0; i<nums.Length; i++) {
            ans[i] = prefix;
            prefix *= nums[i];
        }
        suffix = 1;
        for (int i=nums.Length-1; i>-1; i--) {
            ans[i] *= suffix;
            suffix *= nums[i];
        }
        return ans;
    }

    static void Main(string[] args) {
        int[] nums = {1,2,3,4};
        Console.WriteLine(string.Join(", ", productOfArrayExceptSelf(nums)));
    }
}

File: 2023/Arrays/productOfArrayExceptSelf.py
# Given an integer array nums, return an array answer such that answer[i] is equal to the product of all the elements of nums except nums[i].

# The product of any prefix or suffix of nums is guaranteed to fit in a 32-bit integer.

# You must write an algorithm that runs in O(n) time and without using the division operation.

# Example 1:
# Input: nums = [1,2,3,4]
# Output: [24,12,8,6]

# [1, 1, 2, 6] = LEFT PASS ARRAY 
# [24, 12, 4, 1] = RIGHT PASS ARRAY
# [24, 12, 8, 6] = Final array 

def productExceptSelf(nums):
    ans = [1] * len(nums)
    product = 1
    for i in range(len(nums)):
        ans[i] = product
        product *= nums[i]
    
    product = 1
    for i in range(len(nums)-1, -1, -1):
        ans[i] *= product
        product *= nums[i]
    return ans

print(productExceptSelf([1,2,3,4]))
print(productExceptSelf([-1,1,0,-3,3]))
